File: final_assignment_backend_quarkus/src/main/java/finalassignmentbackend/controller/OffenseTypeController.java
package finalassignmentbackend.controller;

import finalassignmentbackend.entity.OffenseTypeDict;
import finalassignmentbackend.service.OffenseTypeDictService;
import io.smallrye.common.annotation.RunOnVirtualThread;
import jakarta.inject.Inject;
import jakarta.ws.rs.Consumes;
import jakarta.ws.rs.DELETE;
import jakarta.ws.rs.GET;
import jakarta.ws.rs.POST;
import jakarta.ws.rs.PUT;
import jakarta.ws.rs.Path;
import jakarta.ws.rs.PathParam;
import jakarta.ws.rs.Produces;
import jakarta.ws.rs.QueryParam;
import jakarta.ws.rs.HeaderParam;
import jakarta.ws.rs.core.MediaType;
import jakarta.ws.rs.core.Response;
import org.eclipse.microprofile.openapi.annotations.tags.Tag;

import java.util.logging.Level;
import java.util.logging.Logger;

@Path("/api/offense-types")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
@Tag(name = "Offense Type Dictionary", description = "Offense type dictionary management")
public class OffenseTypeController {

    private static final Logger LOG = Logger.getLogger(OffenseTypeController.class.getName());

    @Inject
    OffenseTypeDictService offenseTypeDictService;

    @POST
    @RunOnVirtualThread
    public Response create(OffenseTypeDict request,
                           @HeaderParam("Idempotency-Key") String idempotencyKey) {
        boolean useKey = hasKey(idempotencyKey);
        try {
            if (useKey) {
                if (offenseTypeDictService.shouldSkipProcessing(idempotencyKey)) {
                    return Response.status(208).build();
                }
                offenseTypeDictService.checkAndInsertIdempotency(idempotencyKey, request, "create");
            }
            OffenseTypeDict saved = offenseTypeDictService.createDict(request);
            if (useKey && saved.getTypeId() != null) {
                offenseTypeDictService.markHistorySuccess(idempotencyKey, saved.getTypeId());
            }
            return Response.status(Response.Status.CREATED).entity(saved).build();
        } catch (Exception ex) {
            if (useKey) {
                offenseTypeDictService.markHistoryFailure(idempotencyKey, ex.getMessage());
            }
            LOG.log(Level.SEVERE, "Create offense type failed", ex);
            return Response.status(resolveStatus(ex)).build();
        }
    }

    @PUT
    @Path("/{typeId}")
    @RunOnVirtualThread
    public Response update(@PathParam("typeId") Integer typeId,
                           OffenseTypeDict request,
                           @HeaderParam("Idempotency-Key") String idempotencyKey) {
        boolean useKey = hasKey(idempotencyKey);
        try {
            request.setTypeId(typeId);
            if (useKey) {
                offenseTypeDictService.checkAndInsertIdempotency(idempotencyKey, request, "update");
            }
            OffenseTypeDict updated = offenseTypeDictService.updateDict(request);
            if (useKey && updated.getTypeId() != null) {
                offenseTypeDictService.markHistorySuccess(idempotencyKey, updated.getTypeId());
            }
            return Response.ok(updated).build();
        } catch (Exception ex) {
            if (useKey) {
                offenseTypeDictService.markHistoryFailure(idempotencyKey, ex.getMessage());
            }
            LOG.log(Level.SEVERE, "Update offense type failed", ex);
            return Response.status(resolveStatus(ex)).build();
        }
    }

    @DELETE
    @Path("/{typeId}")
    @RunOnVirtualThread
    public Response delete(@PathParam("typeId") Integer typeId) {
        try {
            offenseTypeDictService.deleteDict(typeId);
            return Response.noContent().build();
        } catch (Exception ex) {
            LOG.log(Level.WARNING, "Delete offense type failed", ex);
            return Response.status(resolveStatus(ex)).build();
        }
    }

    @GET
    @Path("/{typeId}")
    @RunOnVirtualThread
    public Response get(@PathParam("typeId") Integer typeId) {
        try {
            OffenseTypeDict dict = offenseTypeDictService.findById(typeId);
            return dict == null ? Response.status(Response.Status.NOT_FOUND).build() : Response.ok(dict).build();
        } catch (Exception ex) {
            LOG.log(Level.WARNING, "Get offense type failed", ex);
            return Response.status(resolveStatus(ex)).build();
        }
    }

    @GET
    @RunOnVirtualThread
    public Response list() {
        try {
            return Response.ok(offenseTypeDictService.findAll()).build();
        } catch (Exception ex) {
            LOG.log(Level.WARNING, "List offense types failed", ex);
            return Response.status(Response.Status.INTERNAL_SERVER_ERROR).build();
        }
    }

    @GET
    @Path("/search/code/prefix")
    @RunOnVirtualThread
    public Response searchByCodePrefix(@QueryParam("offenseCode") String offenseCode,
                                       @QueryParam("page") Integer page,
                                       @QueryParam("size") Integer size) {
        int resolvedPage = page == null ? 1 : page;
        int resolvedSize = size == null ? 20 : size;
        return Response.ok(offenseTypeDictService.searchByOffenseCodePrefix(offenseCode, resolvedPage, resolvedSize)).build();
    }

    @GET
    @Path("/search/code/fuzzy")
    @RunOnVirtualThread
    public Response searchByCodeFuzzy(@QueryParam("offenseCode") String offenseCode,
                                      @QueryParam("page") Integer page,
                                      @QueryParam("size") Integer size) {
        int resolvedPage = page == null ? 1 : page;
        int resolvedSize = size == null ? 20 : size;
        return Response.ok(offenseTypeDictService.searchByOffenseCodeFuzzy(offenseCode, resolvedPage, resolvedSize)).build();
    }

    @GET
    @Path("/search/name/prefix")
    @RunOnVirtualThread
    public Response searchByNamePrefix(@QueryParam("offenseName") String offenseName,
                                       @QueryParam("page") Integer page,
                                       @QueryParam("size") Integer size) {
        int resolvedPage = page == null ? 1 : page;
        int resolvedSize = size == null ? 20 : size;
        return Response.ok(offenseTypeDictService.searchByOffenseNamePrefix(offenseName, resolvedPage, resolvedSize)).build();
    }

    @GET
    @Path("/search/name/fuzzy")
    @RunOnVirtualThread
    public Response searchByNameFuzzy(@QueryParam("offenseName") String offenseName,
                                      @QueryParam("page") Integer page,
                                      @QueryParam("size") Integer size) {
        int resolvedPage = page == null ? 1 : page;
        int resolvedSize = size == null ? 20 : size;
        return Response.ok(offenseTypeDictService.searchByOffenseNameFuzzy(offenseName, resolvedPage, resolvedSize)).build();
    }

    @GET
    @Path("/search/category")
    @RunOnVirtualThread
    public Response searchByCategory(@QueryParam("category") String category,
                                     @QueryParam("page") Integer page,
                                     @QueryParam("size") Integer size) {
        int resolvedPage = page == null ? 1 : page;
        int resolvedSize = size == null ? 20 : size;
        return Response.ok(offenseTypeDictService.searchByCategory(category, resolvedPage, resolvedSize)).build();
    }

    @GET
    @Path("/search/severity")
    @RunOnVirtualThread
    public Response searchBySeverity(@QueryParam("severityLevel") String severityLevel,
                                     @QueryParam("page") Integer page,
                                     @QueryParam("size") Integer size) {
        int resolvedPage = page == null ? 1 : page;
        int resolvedSize = size == null ? 20 : size;
        return Response.ok(offenseTypeDictService.searchBySeverityLevel(severityLevel, resolvedPage, resolvedSize)).build();
    }

    @GET
    @Path("/search/status")
    @RunOnVirtualThread
    public Response searchByStatus(@QueryParam("status") String status,
                                   @QueryParam("page") Integer page,
                                   @QueryParam("size") Integer size) {
        int resolvedPage = page == null ? 1 : page;
        int resolvedSize = size == null ? 20 : size;
        return Response.ok(offenseTypeDictService.searchByStatus(status, resolvedPage, resolvedSize)).build();
    }

    @GET
    @Path("/search/fine-range")
    @RunOnVirtualThread
    public Response searchByFineRange(@QueryParam("minAmount") double minAmount,
                                      @QueryParam("maxAmount") double maxAmount,
                                      @QueryParam("page") Integer page,
                                      @QueryParam("size") Integer size) {
        int resolvedPage = page == null ? 1 : page;
        int resolvedSize = size == null ? 20 : size;
        return Response.ok(offenseTypeDictService.searchByStandardFineAmountRange(minAmount, maxAmount, resolvedPage, resolvedSize)).build();
    }

    @GET
    @Path("/search/points-range")
    @RunOnVirtualThread
    public Response searchByPointsRange(@QueryParam("minPoints") int minPoints,
                                        @QueryParam("maxPoints") int maxPoints,
                                        @QueryParam("page") Integer page,
                                        @QueryParam("size") Integer size) {
        int resolvedPage = page == null ? 1 : page;
        int resolvedSize = size == null ? 20 : size;
        return Response.ok(offenseTypeDictService.searchByDeductedPointsRange(minPoints, maxPoints, resolvedPage, resolvedSize)).build();
    }

    private boolean hasKey(String value) {
        return value != null && !value.isBlank();
    }

    private Response.Status resolveStatus(Exception ex) {
        return (ex instanceof IllegalArgumentException || ex instanceof IllegalStateException)
                ? Response.Status.BAD_REQUEST
                : Response.Status.INTERNAL_SERVER_ERROR;
    }
}

File: final_assignment_backend_go/project/internal/handler/offense_type_controller.go
package handler

import (
	"errors"
	"net/http"
	"strconv"

	"github.com/gin-gonic/gin"

	"final_assignment_backend_go/project/internal/domain"
	"final_assignment_backend_go/project/internal/service"
)

// OffenseTypeDictServiceContract 处理器侧的违法类型字典服务契约。
type OffenseTypeDictServiceContract interface {
	CheckAndInsertIdempotency(string, *domain.OffenseTypeDict, string) error
	CreateDict(*domain.OffenseTypeDict) (*domain.OffenseTypeDict, error)
	DeleteDict(int) error
	FindAll() ([]domain.OffenseTypeDict, error)
	FindByID(int) (*domain.OffenseTypeDict, error)
	SearchByCategory(string, int, int) ([]domain.OffenseTypeDict, error)
	SearchByDeductedPointsRange(int, int, int, int) ([]domain.OffenseTypeDict, error)
	SearchByOffenseCodeFuzzy(string, int, int) ([]domain.OffenseTypeDict, error)
	SearchByOffenseCodePrefix(string, int, int) ([]domain.OffenseTypeDict, error)
	SearchByOffenseNameFuzzy(string, int, int) ([]domain.OffenseTypeDict, error)
	SearchByOffenseNamePrefix(string, int, int) ([]domain.OffenseTypeDict, error)
	SearchBySeverityLevel(string, int, int) ([]domain.OffenseTypeDict, error)
	SearchByStandardFineAmountRange(float64, float64, int, int) ([]domain.OffenseTypeDict, error)
	SearchByStatus(string, int, int) ([]domain.OffenseTypeDict, error)
	UpdateDict(*domain.OffenseTypeDict) error
}

// OffenseTypeController 对齐 Spring 的 OffenseTypeController（/api/offense-types）。
type OffenseTypeController struct {
	Service OffenseTypeDictServiceContract
}

func NewOffenseTypeController(dictService OffenseTypeDictServiceContract) *OffenseTypeController {
	return &OffenseTypeController{Service: dictService}
}

// RegisterRoutes 注册 /api/offense-types 路由。
func (c *OffenseTypeController) RegisterRoutes(group *gin.RouterGroup) {
	api := group.Group("/offense-types")

	api.POST("", c.create)
	api.GET("", c.list)
	api.GET("/search/code/prefix", c.searchByCodePrefix)
	api.GET("/search/code/fuzzy", c.searchByCodeFuzzy)
	api.GET("/search/name/prefix", c.searchByNamePrefix)
	api.GET("/search/name/fuzzy", c.searchByNameFuzzy)
	api.GET("/search/category", c.searchByCategory)
	api.GET("/search/severity", c.searchBySeverity)
	api.GET("/search/status", c.searchByStatus)
	api.GET("/search/fine-range", c.searchByFineRange)
	api.GET("/search/points-range", c.searchByPointsRange)
	api.PUT("/:typeId", c.update)
	api.DELETE("/:typeId", c.delete)
	api.GET("/:typeId", c.get)
}

// POST /api/offense-types
func (c *OffenseTypeController) create(ctx *gin.Context) {
	var dict domain.OffenseTypeDict
	if err := ctx.ShouldBindJSON(&dict); err != nil {
		ctx.JSON(http.StatusBadRequest, gin.H{"error": "invalid request body"})
		return
	}
	if err := c.Service.CheckAndInsertIdempotency(idempotencyKeyHeader(ctx), &dict, "create"); err != nil {
		if errors.Is(err, service.ErrOffenseTypeDuplicate) {
			ctx.JSON(http.StatusAlreadyReported, gin.H{"success": true, "data": nil})
			return
		}
		ctx.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}
	ctx.JSON(http.StatusCreated, &dict)
}

// PUT /api/offense-types/:typeId
func (c *OffenseTypeController) update(ctx *gin.Context) {
	typeID, err := strconv.Atoi(ctx.Param("typeId"))
	if err != nil || typeID <= 0 {
		ctx.JSON(http.StatusBadRequest, gin.H{"error": "invalid type ID"})
		return
	}
	var dict domain.OffenseTypeDict
	if err := ctx.ShouldBindJSON(&dict); err != nil {
		ctx.JSON(http.StatusBadRequest, gin.H{"error": "invalid request body"})
		return
	}
	dict.TypeID = typeID
	if err := c.Service.CheckAndInsertIdempotency(idempotencyKeyHeader(ctx), &dict, "update"); err != nil {
		if errors.Is(err, service.ErrOffenseTypeDuplicate) {
			ctx.JSON(http.StatusAlreadyReported, gin.H{"success": true, "data": nil})
			return
		}
		ctx.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}
	ctx.JSON(http.StatusOK, &dict)
}

// DELETE /api/offense-types/:typeId
func (c *OffenseTypeController) delete(ctx *gin.Context) {
	typeID, err := strconv.Atoi(ctx.Param("typeId"))
	if err != nil || typeID <= 0 {
		ctx.JSON(http.StatusBadRequest, gin.H{"error": "invalid type ID"})
		return
	}
	if err := c.Service.DeleteDict(typeID); err != nil {
		ctx.JSON(http.StatusInternalServerError, gin.H{"error": "failed to delete offense type"})
		return
	}
	ctx.Status(http.StatusNoContent)
}

// GET /api/offense-types/:typeId
func (c *OffenseTypeController) get(ctx *gin.Context) {
	typeID, err := strconv.Atoi(ctx.Param("typeId"))
	if err != nil || typeID <= 0 {
		ctx.JSON(http.StatusBadRequest, gin.H{"error": "invalid type ID"})
		return
	}
	dict, err := c.Service.FindByID(typeID)
	if err != nil {
		ctx.JSON(http.StatusNotFound, gin.H{"error": "offense type not found"})
		return
	}
	ctx.JSON(http.StatusOK, dict)
}

// GET /api/offense-types
func (c *OffenseTypeController) list(ctx *gin.Context) {
	dicts, err := c.Service.FindAll()
	if err != nil {
		ctx.JSON(http.StatusInternalServerError, gin.H{"error": "failed to list offense types"})
		return
	}
	ctx.JSON(http.StatusOK, dicts)
}

// GET /api/offense-types/search/code/prefix?offenseCode=
func (c *OffenseTypeController) searchByCodePrefix(ctx *gin.Context) {
	page, size := paginationParams(ctx)
	dicts, err := c.Service.SearchByOffenseCodePrefix(ctx.Query("offenseCode"), page, size)
	dictResult(ctx, dicts, err)
}

// GET /api/offense-types/search/code/fuzzy?offenseCode=
func (c *OffenseTypeController) searchByCodeFuzzy(ctx *gin.Context) {
	page, size := paginationParams(ctx)
	dicts, err := c.Service.SearchByOffenseCodeFuzzy(ctx.Query("offenseCode"), page, size)
	dictResult(ctx, dicts, err)
}

// GET /api/offense-types/search/name/prefix?offenseName=
func (c *OffenseTypeController) searchByNamePrefix(ctx *gin.Context) {
	page, size := paginationParams(ctx)
	dicts, err := c.Service.SearchByOffenseNamePrefix(ctx.Query("offenseName"), page, size)
	dictResult(ctx, dicts, err)
}

// GET /api/offense-types/search/name/fuzzy?offenseName=
func (c *OffenseTypeController) searchByNameFuzzy(ctx *gin.Context) {
	page, size := paginationParams(ctx)
	dicts, err := c.Service.SearchByOffenseNameFuzzy(ctx.Query("offenseName"), page, size)
	dictResult(ctx, dicts, err)
}

// GET /api/offense-types/search/category?category=
func (c *OffenseTypeController) searchByCategory(ctx *gin.Context) {
	page, size := paginationParams(ctx)
	dicts, err := c.Service.SearchByCategory(ctx.Query("category"), page, size)
	dictResult(ctx, dicts, err)
}

// GET /api/offense-types/search/severity?severityLevel=
func (c *OffenseTypeController) searchBySeverity(ctx *gin.Context) {
	page, size := paginationParams(ctx)
	dicts, err := c.Service.SearchBySeverityLevel(ctx.Query("severityLevel"), page, size)
	dictResult(ctx, dicts, err)
}

// GET /api/offense-types/search/status?status=
func (c *OffenseTypeController) searchByStatus(ctx *gin.Context) {
	page, size := paginationParams(ctx)
	dicts, err := c.Service.SearchByStatus(ctx.Query("status"), page, size)
	dictResult(ctx, dicts, err)
}

// GET /api/offense-types/search/fine-range?minAmount=&maxAmount=
func (c *OffenseTypeController) searchByFineRange(ctx *gin.Context) {
	page, size := paginationParams(ctx)
	minAmount, errMin := strconv.ParseFloat(ctx.Query("minAmount"), 64)
	maxAmount, errMax := strconv.ParseFloat(ctx.Query("maxAmount"), 64)
	if errMin != nil || errMax != nil {
		ctx.JSON(http.StatusBadRequest, gin.H{"error": "minAmount and maxAmount must be numbers"})
		return
	}
	dicts, err := c.Service.SearchByStandardFineAmountRange(minAmount, maxAmount, page, size)
	dictResult(ctx, dicts, err)
}

// GET /api/offense-types/search/points-range?minPoints=&maxPoints=
func (c *OffenseTypeController) searchByPointsRange(ctx *gin.Context) {
	page, size := paginationParams(ctx)
	minPoints, errMin := strconv.Atoi(ctx.Query("minPoints"))
	maxPoints, errMax := strconv.Atoi(ctx.Query("maxPoints"))
	if errMin != nil || errMax != nil {
		ctx.JSON(http.StatusBadRequest, gin.H{"error": "minPoints and maxPoints must be integers"})
		return
	}
	dicts, err := c.Service.SearchByDeductedPointsRange(minPoints, maxPoints, page, size)
	dictResult(ctx, dicts, err)
}

func dictResult(ctx *gin.Context, dicts []domain.OffenseTypeDict, err error) {
	if err != nil {
		ctx.JSON(http.StatusInternalServerError, gin.H{"error": "query failed"})
		return
	}
	ctx.JSON(http.StatusOK, dicts)
}
